src/components/common/ClusteringAddPopup.tsx

```diff
@@ -0,0 +1,288 @@
+import React, { Component } from 'react';
+import {
+    Icon,
+    Popup,
+    Button,
+    Modal,
+    Divider,
+    Header,
+    Card,
+    CardContent,
+    Form,
+    Input,
+} from 'semantic-ui-react';
+import { BackendAPI } from './API';
+import * as R from 'ramda';
+
+import CSVReader from 'react-csv-reader';
+
+declare const DEBUG: boolean;
+
+interface ClusteringAddPopupProps {
+    handleModalOpen: Function;
+    handleModalClose: Function;
+    orcid_name: string;
+    orcid_id: string;
+    orcid_uuid: string;
+}
+
+interface ClusteringAddPopupState {
+    status: string;
+    showModal: boolean;
+    clusteringName: string;
+    newclusterInfo?: {
+        cellIDs: string[];
+        clusterIDs: string[];
+    };
+}
+
+export default class ClusteringAddPopup extends Component<
+    ClusteringAddPopupProps,
+    ClusteringAddPopupState
+> {
+    constructor(props) {
+        super(props);
+        this.state = {
+            status: 'ready',
+            showModal: false,
+            clusteringName: '',
+        };
+    }
+
+    closeModal = () => {
+        this.setState({ showModal: false });
+        this.props.handleModalClose();
+    };
+
+    handleModalOpen = () => {
+        this.props.handleModalOpen();
+    };
+
+    onClusteringNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        this.setState({ clusteringName: e.target.value });
+    };
+
+    validateClusterData = (data: [Array<string>, Array<string>]) => {
+        let cols = new Set();
+        data.map((x) => {
+            cols.add(x.length);
+        });
+
+        cols.delete(0);
+        if (cols.size != 1) {
+            this.setState({ status: 'dataError' });
+            console.log(cols);
+            alert(
+                'Mismatched column lengths, please select a correctly formatted file!'
+            );
+        } else if (!cols.has(2)) {
```

Comment:
This doesn't do what you think it does, as the data is still `n`x`2`, before the zip, not `2`x`n`. That's the purpose of the set/map above, to ensure we have consistency for entry sizes and that we have 2 columns per entry

Reply:
I'm trying to prove that our assertion on line 84 (below) about the shape of `data` is correct. I can't make sense of your shape notation (column or row major order?). The fundamental point is that on line 84, we assert that the outer array contains 2 elements. By the time we get to the `else` on line 81 this needs to be proven true.

Reply:
The data comes into the function as row-major order, which mean `data.length` will report the number of rows. The two ifs on 71 and 76 will assert that our shape is true, because we calculate the number of columns per row (from the row-major data), and confirm we only have unique 1 value (if on 71), then that value must equal 2 (if on 76).